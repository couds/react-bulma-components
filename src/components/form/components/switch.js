import cn from 'classnames';
import PropTypes from 'prop-types';
import React, { useState } from 'react';
import CONSTANTS from '../../../constants';
import modifiers from '../../../modifiers';

const colors = Object.values(CONSTANTS.COLORS);

export const Switch = ({ className, style, disabled, value, children, checked, color, size, name, ...allProps }) => {
  const props = modifiers.clean(allProps);
  const [isMouseDown, setMouseDown] = useState(false);
  return (
    <label
      className={cn('switch', modifiers.getClassName(allProps), className)}
      style={style}
      onMouseDown={() => setMouseDown(true)}
      onMouseUp={() => setMouseDown(false)}
      onMouseOut={() => setMouseDown(false)}
      disabled={disabled}
    >
      <input {...props} name={name} type="checkbox" value={value} disabled={disabled} checked={checked} />
      <span
        className={cn('check', {
          'is-elastic': isMouseDown && !disabled,
          [`is-${color}`]: color,
          [`is-${size}`]: size
        })}
        data-testid="switch-check"
      />
      <span className="control-label">{children}</span>
    </label>
  );
};

Switch.propTypes = {
  ...modifiers.propTypes,
  children: PropTypes.node,
  className: PropTypes.string,
  style: PropTypes.shape({}),
  color: PropTypes.oneOf(colors),
  size: PropTypes.oneOf(Object.values(CONSTANTS.SIZES)),
  disabled: PropTypes.bool,
  value: PropTypes.string,
  checked: PropTypes.bool,
  name: PropTypes.string
};

Switch.defaultProps = {
  ...modifiers.defaultProps,
  children: null,
  className: '',
  value: '',
  style: {},
  color: null,
  size: null,
  disabled: false,
  checked: false,
  name: null
};
