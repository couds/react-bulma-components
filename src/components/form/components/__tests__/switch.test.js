import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import { Switch } from '../switch';

describe('Switch component', () => {
  it('should render', () => {
    const { asFragment } = render(<Switch>Text</Switch>);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should change value on change event', () => {
    const spy = jest.fn();
    const { getByLabelText } = render(<Switch onChange={spy}>Text</Switch>);
    fireEvent.click(getByLabelText(/Text/i));
    expect(spy).toHaveBeenCalledTimes(1);
  });
  it('should have info color classname', () => {
    const { asFragment } = render(<Switch color="info">Text</Switch>);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should have medium classname', () => {
    const { asFragment } = render(<Switch size="medium">Text</Switch>);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should be disabled', () => {
    const { asFragment } = render(<Switch disabled>Text</Switch>);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should be elastic', () => {
    const { getByLabelText, getByTestId } = render(<Switch>Text</Switch>);
    fireEvent.mouseDown(getByLabelText(/Text/i));
    expect(getByTestId('switch-check')).toHaveClass('is-elastic');
  });
  it('should not be elastic', () => {
    const { getByLabelText, getByTestId } = render(<Switch>Text</Switch>);
    fireEvent.mouseOut(getByLabelText(/Text/i));
    expect(getByTestId('switch-check')).not.toHaveClass('is-elastic');
  });
  it('should not be elastic 2', () => {
    const { getByLabelText, getByTestId } = render(<Switch>Text</Switch>);
    fireEvent.mouseUp(getByLabelText(/Text/i));
    expect(getByTestId('switch-check')).not.toHaveClass('is-elastic');
  });
  it('should set input checked if checked', () => {
    const { getByTestId } = render(<Switch checked />);
    expect(getByTestId('switch-input')).toHaveAttribute('checked', '');
  });
});
