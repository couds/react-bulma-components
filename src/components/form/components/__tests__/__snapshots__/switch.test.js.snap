// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch component should be disabled 1`] = `
<DocumentFragment>
  <label
    class="switch"
    data-testid="switch-label"
    disabled=""
  >
    <input
      data-testid="switch-input"
      disabled=""
      type="checkbox"
      value=""
    />
    <span
      class="check"
      data-testid="switch-check"
    />
    <span
      class="control-label"
    >
      Text
    </span>
  </label>
</DocumentFragment>
`;

exports[`Switch component should have info color classname 1`] = `
<DocumentFragment>
  <label
    class="switch"
    data-testid="switch-label"
  >
    <input
      data-testid="switch-input"
      type="checkbox"
      value=""
    />
    <span
      class="check is-info"
      data-testid="switch-check"
    />
    <span
      class="control-label"
    >
      Text
    </span>
  </label>
</DocumentFragment>
`;

exports[`Switch component should have medium classname 1`] = `
<DocumentFragment>
  <label
    class="switch"
    data-testid="switch-label"
  >
    <input
      data-testid="switch-input"
      type="checkbox"
      value=""
    />
    <span
      class="check is-medium"
      data-testid="switch-check"
    />
    <span
      class="control-label"
    >
      Text
    </span>
  </label>
</DocumentFragment>
`;

exports[`Switch component should render 1`] = `
<DocumentFragment>
  <label
    class="switch"
    data-testid="switch-label"
  >
    <input
      data-testid="switch-input"
      type="checkbox"
      value=""
    />
    <span
      class="check"
      data-testid="switch-check"
    />
    <span
      class="control-label"
    >
      Text
    </span>
  </label>
</DocumentFragment>
`;
